File: src/main/java/ch/mse/app/types/CompoundTypes.java
package ch.mse.app.types;

public class CompoundTypes {
    class Component{}
    class Point{}

    interface Container { // controls graphical objects (components)
        void add(Component c, Point pos);
        void foo(Container c);
    }

    interface TextEditor { // interface for a text editor
        void insert(char ch, int pos);
        Point coordinates(int pos);
        void foo(TextEditor c);
    }

    public <T extends TextEditor & Container> void insertComponent(
                                            T obj, Component c, int pos) {
        var point = obj.coordinates(pos);
        obj.add(c, point);

        obj.foo((Container)obj);
    }

}

File: src/main/scala/ch/mse/app/traits/CompoundTypes.scala
package ch.mse.app.traits

object CompoundTypes {
    trait Component
    trait Point

    trait Container { // controls graphical objects (components)
        def add(c: Component, pos: Point) : Unit
    }

    trait TextEditor { // trait for a text editor
        def insert(ch: Char, pos: Int): Unit
        def coordinates(pos: Int) : Point
    }

    def insertComponent(obj: TextEditor & Container, c: Component, pos: Int): Unit = {
        val point = obj.coordinates(pos) // obj <: TextEditor
        obj.add(c, point)                // obj <: Container
    }

}
